File: files/fraser-coho-workshop/scripts/review_ai.R
# Optional live Chapter 5 exercise. No purchase, paid model or paid fallback.
# Run from the extracted project root after completing Chapters 2-4.
library(metasalmon)
source("scripts/workshop.R")
check_preparation()
if (!nzchar(Sys.getenv("OPENROUTER_API_KEY"))) stop("No free API key configured. Use ai/README.md and the recorded assessments instead.")
pkg <- read_salmon_datapackage(workshop_path)
dict <- pkg$dictionary[pkg$dictionary$column_name == "NATURAL_ADULT_SPAWNERS", , drop = FALSE]
for (field in c("term_iri", "property_iri", "entity_iri", "unit_iri", "constraint_iri", "statistical_modifier_iri")) dict[[field]] <- NA_character_
context <- file.path("worksheets", c("dataset-nodes.csv", "dataset-edges.csv", "data-dictionary.csv", "variable-decomposition.csv"))
out <- file.path("output", paste0("ai-live-r-", format(Sys.time(), "%Y%m%dT%H%M%S", tz = "UTC")))
if (dir.exists(out)) stop("Output exists; wait a second before starting a separate attempt.")
dir.create(out, recursive = TRUE)
records <- list()
attempts <- 0L

# The public callback seam lets this teaching script count actual HTTP attempts
# and retain request/response evidence. It does not change package APIs.
# Retire the callback when the released package exposes equivalent evidence
# and request-budget controls. Never log configuration or authorization headers.
free_request <- function(messages, config) {
  if (!identical(config$provider, "openrouter") || !identical(config$model, "openrouter/free")) stop("Only openrouter/free is permitted by this exercise.")
  if (attempts >= 8L) stop("Workshop request limit reached. Continue with recorded assessments.")
  attempts <<- attempts + 1L
  body <- list(model = "openrouter/free", messages = messages,
               temperature = 0, max_tokens = 2500,
               response_format = list(type = "json_object"))
  # No HTTP-level automatic retries. Any package retry re-enters this callback
  # and counts against the same maximum of eight requests.
  request <- httr2::request("https://openrouter.ai/api/v1/chat/completions") |>
    httr2::req_headers(Authorization = paste("Bearer", Sys.getenv("OPENROUTER_API_KEY"))) |>
    httr2::req_body_json(body, auto_unbox = TRUE) |>
    httr2::req_timeout(45) |>
    httr2::req_error(is_error = function(resp) FALSE)
  response <- tryCatch(httr2::req_perform(request), error = function(e) NULL)
  status <- if (is.null(response)) 0L else httr2::resp_status(response)
  payload <- if (is.null(response)) NULL else tryCatch(httr2::resp_body_json(response, simplifyVector = FALSE), error = function(e) NULL)
  # Only successful content is retained; arbitrary provider error bodies can
  # contain request diagnostics. Status codes are sufficient for failure review.
  record <- list(attempt = attempts, utc = format(Sys.time(), tz = "UTC", usetz = TRUE),
                 request = body, http_status = status)
  if (status == 200L) record$response <- payload
  records[[length(records) + 1L]] <<- record
  jsonlite::write_json(records, file.path(out, "requests.json"), auto_unbox = TRUE, pretty = TRUE)
  if (status != 200L) stop("Free model HTTP request failed, status ", status, ". Use recorded assessments if unavailable.")
  content <- payload$choices[[1]]$message$content
  if (is.null(content) || !nzchar(content)) stop("Free model returned no assessment content.")
  content <- sub("^```(?:json)?[[:space:]]*", "", trimws(content), perl = TRUE)
  content <- sub("[[:space:]]*```$", "", content, perl = TRUE)
  jsonlite::fromJSON(content, simplifyVector = FALSE)
}

result <- suggest_semantics(
  df = pkg$resources, dict = dict, sources = c("smn", "gcdfo"),
  max_per_role = 3, llm_assess = TRUE, llm_provider = "openrouter",
  llm_model = "openrouter/free", llm_top_n = 3,
  llm_context_files = context, llm_request_fn = free_request
)
suggestions <- semantic_suggestions(result)
assessments <- semantic_llm_assessments(result)
successful <- if (is.null(assessments) || !nrow(assessments)) 0L else sum(
  !is.na(assessments$llm_decision) & nzchar(assessments$llm_decision) &
    (is.na(assessments$llm_error) | !nzchar(assessments$llm_error)))
readr::write_csv(suggestions, file.path(out, "suggestions.csv"), na = "")
readr::write_csv(assessments, file.path(out, "assessments.csv"), na = "")
hashes <- stats::setNames(vapply(context, function(p) digest::digest(file = p, algo = "sha256", serialize = FALSE), character(1)), context)
jsonlite::write_json(list(
  kind = "live-provider-recording", package = "metasalmon", version = as.character(packageVersion("metasalmon")),
  requested_model = "openrouter/free", actual_http_attempts = attempts,
  successful_assessment_rows = successful,
  rehearsal_status = if (successful > 0L) "response-recorded-needs-human-review" else "failed-no-successful-assessment",
  context_sha256 = as.list(hashes), source_csv_sha256 = digest::digest(file = workshop_source, algo = "sha256", serialize = FALSE),
  note = "Provider responses and package decisions are proposals, not human approval. Actual routed model IDs are in requests.json when returned."
), file.path(out, "provenance.json"), auto_unbox = TRUE, pretty = TRUE)
message("Saved ", attempts, " HTTP attempts and ", successful, " successful assessment rows at ", out, ".")
if (successful == 0L) stop("No successful AI assessment was returned. Retain the failure receipt and use the recorded comparison activity.")
message("Compare the recorded response against the human model before any application.")

File: files/fraser-coho-workshop/scripts/review_ai.py
"""Optional free-only live Chapter 5 exercise; no paid fallback or purchase."""
import datetime
import hashlib
import json
import os
import re
from pathlib import Path
import urllib.error
import urllib.request
import metasalmonpy as ms
from workshop import PACKAGE, SOURCE, check_preparation

check_preparation()
if not os.environ.get("OPENROUTER_API_KEY"):
    raise SystemExit("No free key configured. Use ai/README.md and recorded assessments instead.")
pkg = ms.read_salmon_datapackage(str(PACKAGE))
dictionary = pkg["dictionary"].loc[pkg["dictionary"].column_name == "NATURAL_ADULT_SPAWNERS"].copy()
for field in ("term_iri", "property_iri", "entity_iri", "unit_iri", "constraint_iri", "statistical_modifier_iri"):
    dictionary[field] = ""
context = [str(Path("worksheets") / name) for name in ("dataset-nodes.csv", "dataset-edges.csv", "data-dictionary.csv", "variable-decomposition.csv")]
out = Path("output") / ("ai-live-python-" + datetime.datetime.now(datetime.timezone.utc).strftime("%Y%m%dT%H%M%S"))
out.mkdir(parents=True, exist_ok=False)
records = []
attempts = 0


def free_request(messages, config):
    global attempts
    if config["provider"] != "openrouter" or config["model"] != "openrouter/free":
        raise ValueError("Only openrouter/free is permitted.")
    if attempts >= 8:
        raise ValueError("Workshop request limit reached; continue with recorded assessments.")
    attempts += 1
    body = {"model": "openrouter/free", "messages": messages, "temperature": 0,
            "max_tokens": 2500, "response_format": {"type": "json_object"}}
    req = urllib.request.Request("https://openrouter.ai/api/v1/chat/completions",
        data=json.dumps(body).encode(), headers={"Content-Type": "application/json",
        "Authorization": "Bearer " + os.environ["OPENROUTER_API_KEY"]})
    payload = None
    status = 0
    parse_error = False
    try:
        with urllib.request.urlopen(req, timeout=45) as response:
            status = response.status
            try:
                payload = json.load(response)
            except (ValueError, UnicodeDecodeError):
                parse_error = True
    except urllib.error.HTTPError as error:
        status = error.code
    except (urllib.error.URLError, TimeoutError):
        status = 0
    record = {"attempt": attempts, "utc": datetime.datetime.now(datetime.timezone.utc).isoformat(),
              "request": body, "http_status": status, "parse_error": parse_error}
    if status == 200:
        record["response"] = payload
    records.append(record)
    (out / "requests.json").write_text(json.dumps(records, indent=2) + "\n")
    if status != 200:
        raise RuntimeError(f"Free model request failed, HTTP {status}. Use saved assessments if unavailable.")
    if parse_error or payload is None:
        raise ValueError("Free model returned invalid response JSON; request was counted.")
    content = payload["choices"][0]["message"].get("content") or ""
    content = re.sub(r"^```(?:json)?\s*", "", content.strip())
    content = re.sub(r"\s*```$", "", content)
    return json.loads(content)


result = ms.suggest_semantics(df=pkg["resources"], dict_df=dictionary, sources=["smn", "gcdfo"],
    max_per_role=3, llm_assess=True, llm_provider="openrouter", llm_model="openrouter/free",
    llm_api_key=os.environ["OPENROUTER_API_KEY"], llm_top_n=3, llm_context_files=context,
    llm_request_fn=free_request)
for attribute, filename in (("semantic_suggestions", "suggestions.csv"), ("semantic_llm_assessments", "assessments.csv")):
    evidence = result.attrs.get(attribute)
    if evidence is not None:
        evidence.to_csv(out / filename, index=False)
assessments = result.attrs.get("semantic_llm_assessments")
successful = 0 if assessments is None or assessments.empty else int((
    assessments["llm_decision"].fillna("").ne("") & assessments["llm_error"].fillna("").eq("")
).sum())
(out / "provenance.json").write_text(json.dumps({
    "kind": "live-provider-recording", "package": "metasalmonpy", "version": ms.__version__,
    "requested_model": "openrouter/free", "actual_http_attempts": attempts,
    "successful_assessment_rows": successful,
    "rehearsal_status": "response-recorded-needs-human-review" if successful else "failed-no-successful-assessment",
    "context_sha256": {p: hashlib.sha256(Path(p).read_bytes()).hexdigest() for p in context},
    "source_csv_sha256": hashlib.sha256(SOURCE.read_bytes()).hexdigest(),
    "note": "Provider responses and package decisions are proposals, not human approval. Actual routed model IDs are in requests.json when returned."
}, indent=2) + "\n")
print(f"Saved {attempts} HTTP attempts and {successful} successful assessment rows at {out}.")
if not successful:
    raise SystemExit("No successful AI assessment was returned. Retain the failure receipt and use the recorded comparison activity.")
print("Compare the recorded response against the human model before any application.")
